beamer/agent/middleware.py

```diff
@@ -21,15 +25,29 @@ def _result_ok(response: RPCResponse) -> bool:
     return response.get("result") is not None
 
 
+def cache_get_block_by_number(
+    make_request: Callable[[RPCEndpoint, Any], RPCResponse], _w3: Web3, chain_id: ChainId
+) -> Callable[[RPCEndpoint, Any], RPCResponse]:
+    global _BLOCK_CACHE
+    global _LATEST_BLOCK_NUMBER_MAP
+
+    if chain_id not in _BLOCK_CACHE:
+        _BLOCK_CACHE[chain_id] = lru.LRU(1000)
+    if chain_id not in _LATEST_BLOCK_NUMBER_MAP:
+        _LATEST_BLOCK_NUMBER_MAP[chain_id] = None
+    return _cache_get_block_by_number(make_request, _w3, chain_id)
+
+
 # This middleware caches responses of eth_getBlockByNumber calls. Requests with
 # block number set to 'latest' will never be returned from the cache. However,
 # successful responses of those calls will be stored into the cache so that
 # further requests specifying a concrete block number can be satisfied from
 # the cache.
```

Comment:
Nitpick: I'd move this comment to above the public entry point, `cache_get_block_by_number`, the function below is an implementation detail.